src/Pact/Analyze/Types.hs

```diff
@@ -646,27 +541,146 @@ data Prop a where
+pattern Inj :: sub :<: sup => sub a -> sup a
+pattern Inj a <- (project -> Just a) where
+  Inj a = inject a
+
+instance PropSpecific :<: Prop where
+  inject = PropSpecific
+  project = \case
+    PropSpecific a -> Just a
+    _              -> Nothing
+
+instance Numerical Prop :<: Prop where
+  inject = PureProp . Numerical
+  project (PureProp (Numerical a)) = Just a
+  project _                        = Nothing
+
+instance Numerical Invariant :<: Invariant where
+  inject = PureInvariant . Numerical
+  project (PureInvariant (Numerical a)) = Just a
+  project _                             = Nothing
+
+data PureTerm et t a where
+  Lit :: a -> PureTerm et t a
+  -- | Injects a symbolic value into the language
+  Sym :: S a -> PureTerm et t a
+
+  -- string ops
+  -- | The concatenation of two 'String' expressions
+  StrConcat :: t String -> t String -> PureTerm et t String
+  -- | The length of a 'String' expression
+  StrLength :: t String                     -> PureTerm et t Integer
+
+  -- numeric ops
+  Numerical :: Numerical t a -> PureTerm et t a
+
+  -- Time
+  -- | Adds an 'Integer' expression to a 'Time' expression
+  IntAddTime      :: t Time -> t Integer -> PureTerm et t Time
+  -- | Adds a 'Decimal' expression to a 'Time' expression
+  DecAddTime      :: t Time -> t Decimal -> PureTerm et t Time
+
+  -- comparison
+  -- | A 'ComparisonOp' expression over two 'Integer' expressions
+  IntegerComparison :: ComparisonOp -> t Integer -> t Integer -> PureTerm et t Bool
+  -- | A 'ComparisonOp' expression over two 'Decimal' expressions
+  DecimalComparison :: ComparisonOp -> t Decimal -> t Decimal -> PureTerm et t Bool
+  -- | A 'ComparisonOp' expression over two 'Time' expressions
+  TimeComparison    :: ComparisonOp -> t Time    -> t Time    -> PureTerm et t Bool
+  -- | A 'ComparisonOp' expression over two 'String' expressions
+  StringComparison  :: ComparisonOp -> t String  -> t String  -> PureTerm et t Bool
+  -- | A 'ComparisonOp' expression over two 'Bool' expressions
+  BoolComparison    :: ComparisonOp -> t Bool    -> t Bool    -> PureTerm et t Bool
+
+  KeySetEqNeq :: EqNeq -> t KeySet -> t KeySet -> PureTerm et t Bool
+  ObjectEqNeq :: EqNeq -> t Object -> t Object -> PureTerm et t Bool
+
+  At             :: Schema -> t String -> t Object -> EType -> PureTerm et t a
+
+  LiteralObject  :: Map Text et -> PureTerm et t Object
+
+  -- boolean ops
+  -- | A 'Logical' expression over one or two 'Bool' expressions; one operand
+  -- for NOT, and two operands for AND or OR.
+  Logical :: LogicalOp -> [t Bool] -> PureTerm et t Bool
+
+deriving instance Show a => Show (PureTerm EProp Prop a)
+deriving instance Show a => Show (PureTerm EInvariant Invariant a)
+
+deriving instance Eq a => Eq (PureTerm EProp Prop a)
+deriving instance Eq a => Eq (PureTerm EInvariant Invariant a)
+
+pattern PLit :: a -> Prop a
+pattern PLit a = PureProp (Lit a)
+
+data Prop a where
+  PropSpecific :: PropSpecific a  -> Prop a
```

Comment:
`Prop` is one style of term where we put everything that's not pure in its own data type, `PropSpecific`. Check out `Term` for the other style. I'm not sure yet which I prefer.